### PR TITLE
Add support for Zitadel authentication

### PR DESCRIPTION
**Pull request by thepaperpilot**
This PR adds Zitadel, a self hostable identity provider, support for OAuth logins. It works similarly to the keycloak one, but with zitadel URIs instead. It has a few more supporting files due to there not being an existing package for zitadel in the oauth bundle mbin uses.

### CODE REVIEW

File: src/Security/ZitadelAuthenticator.php
Comment (e-five256) [inline src/Security/ZitadelAuthenticator.php:98]: should this be:

```suggestion
                $user->oauthZitadelId = $zitadelUser->getId();
```

File: src/Security/ZitadelAuthenticator.php
Comment (e-five256) [inline src/Security/ZitadelAuthenticator.php:16]: I sort of would've expected a composer vendor change for this, but I can't say for sure I know how either keycloak or zitadel work. Keycloak seems to pull in https://github.com/stevenmaguire/oauth2-keycloak/blob/master/src/Provider/KeycloakResourceOwner.php and adds it as a composer dependency

Comment (thepaperpilot) [reply src/Security/ZitadelAuthenticator.php:16]: Pushing a commit for PR feedback soon, but just wanted to point out that was actually the wrong path, from when I was still figuring stuff out. It'll be updated to App\Provider\ZitadelResourceOwner, as the file is within the repository itself.

**Review by asdfzdfj — COMMENTED** [review]
It'd be appreciated if you could provide some description for the PR

File: src/Security/ZitadelAuthenticator.php
Comment (thepaperpilot) [reply src/Security/ZitadelAuthenticator.php:83]: On my instance the `username` definitely exists, but `preferred_username` included the domain which isn't what I wanted (e.g. I'd get `thepaperpilot_at_domain.tld`)

Comment (thepaperpilot) [reply src/Security/ZitadelAuthenticator.php:83]: On [this page](https://zitadel.com/docs/apis/openidoauth/claims) in the docs it lists what fields are provided under which requests, and it has a row indicating it's provided when requested in the introspect request if the scope is specifically requested (the third column), which it is as part of the `profile` scope.
image

Comment (asdfzdfj) [reply src/Security/ZitadelAuthenticator.php:83]: hmm ok, it might as well be zitadel misconfiguration on my part. for the `username` containing the `-at-` part is probably because zitadel likes usernames in that format, but it looks like when creating a user via admin, you can just put in user name without `@domain.tld` part and it'll accept it just fine, and that's what I did when creating a test user for testing this patch, guess I'll just let you know that this can happen for possible investigation 

for the record, this is what I get from dumping the `$zitadelUser->toArray()` just before getting username

```php
$username = $slugger->slug(dump($zitadelUser->toArray())['username'])
```

```php
 array:9 [▼
  "sub" => "249646190008795139"
  "name" => "ZITADEL Admin"
  "given_name" => "ZITADEL"
  "family_name" => "Admin"
  "locale" => "en"
  "updated_at" => 1705296726
  "preferred_username" => "tama@zitadel.localhost"
  "email" => "tama@zitadel.localhost"
  "email_verified" => true
]
```

Comment (thepaperpilot) [reply src/Security/ZitadelAuthenticator.php:83]: I found out you were right and my code was actually preventing new users from authenticating

**Comment by thepaperpilot** [comment]
Before this gets merged, I'd like to update the authenticator to update the user's username and avatar etc.

**Comment by e-five256** [comment]
there's a button to convert PRs to drafts on the sidebar, we've been using that to mark/unmark things as still in progress / ready to merge, definitely recommend it to avoid us merging too early accidentally

image

**Comment by thepaperpilot** [comment]
Sorry for the delay - it now copies over the avatar though.

File: .env.example
Comment (melroy89) [inline .env.example:73]: does Mbin keeps working without adding this to the config?

Comment (thepaperpilot) [reply .env.example:73]: It should, but since that file is supposed to show the possible keys I think its fine

**Comment by BentiGorlich** [comment]
@thepaperpilot what is the current status? Are you finished, and only waiting for a review?

**Comment by thepaperpilot** [comment]
> @thepaperpilot what is the current status? Are you finished, and only waiting for a review?

That is correct

**Comment by BentiGorlich** [comment]
Ok, Then I will put that on my todo list :+1:

File: src/Security/ZitadelAuthenticator.php
Comment (BentiGorlich) [inline src/Security/ZitadelAuthenticator.php:148]: Could you refactor this, so that it is not duplicated in 3 authenticators (Zitadel, Google and Facebook)

File: templates/components/login_socials.html.twig
Comment (BentiGorlich) [inline templates/components/login_socials.html.twig:21]: Could we get the icon of the zitadel org here? If it is too much effort that is fine. Maybe the Zitadel Logo itself (if they have a monochrome svg) would suffice?

**Review by BentiGorlich — COMMENTED** [review]
If both remarks are too much work this would be good to go. 
It works and has the same standard as the other authenticators. Good job @thepaperpilot

File: templates/components/login_socials.html.twig
Comment (thepaperpilot) [reply templates/components/login_socials.html.twig:21]: I don't think that's plausible. I checked my own instance and the url is `https://auth.incremental.social/ui/login/resources/dynamic?orgId=231964622825979915&default-policy=true&filename=policy/label/logo-dark-251434450057953312`. I have no idea how I'd get that org id though. That said, I'd urge against using the zitadel logo itself since typically instances are going to be using custom branding. In fact, I'm thinking perhaps the best thing to do would be adding 2 more env vars for the display name and logo to use for zitadel auth

File: .env.example
Comment (melroy89) [reply .env.example:73]: so what I mean, is most current admin / server owners will forget to add this configuration. So Mbin should still work fine, even if `OAUTH_ZITADEL_ID` is undefined (not present in the `.env` file).

File: templates/components/login_socials.html.twig
Comment (melroy89) [reply templates/components/login_socials.html.twig:21]: Of course it's possible. But currently we seem to use font-awesome icons (`fa`).  So now it's showing the `fa-lock` icon. 

However, font awesome doesn't have an icon / logo for Zitadel it seems. The icon seems to be come kind of triangle. so maybe `fa-caret-left` would do it for now?

image

Comment (BentiGorlich) [reply templates/components/login_socials.html.twig:21]: > That said, I'd urge against using the zitadel logo itself since typically instances are going to be using custom branding. In fact, I'm thinking perhaps the best thing to do would be adding 2 more env vars for the display name and logo to use for zitadel auth

I am fine with either just keeping the lock icon or the 2 env vars with logo and name. :shrug:

Comment (melroy89) [reply templates/components/login_socials.html.twig:21]: But from the user perspective it makes sense that you know you are going to Zitadel. Right?

I don't care if Zitadel then support custom logins /logos or alike. But I digress

File: src/Security/ZitadelAuthenticator.php
Comment (e-five256) [reply src/Security/ZitadelAuthenticator.php:148]: I don't mind taking a shot at that when I go in to add the disable sso registration stuff as a follow up

File: templates/components/login_socials.html.twig
Comment (e-five256) [reply templates/components/login_socials.html.twig:21]: I think for a first pass as is this is fine, thepaperpilot is the main user of it right now and likely already can write bespoke stuff if they haven't already. can always make it more customizable in a followup

**Review by e-five256 — APPROVED** [review]
seemed good to me, had a bit of trouble with zitadel itself in my own dev env, but obviously thepaperpilot is running this and I assume not seeing issues

File: templates/components/login_socials.html.twig
Comment (BentiGorlich) [reply templates/components/login_socials.html.twig:21]: Sounds good to me

**Review by BentiGorlich — APPROVED** [review]

**Pull request merged**
